File: Oracle/worker.py
import time
import os
import aleo
import requests
from dotenv import load_dotenv
import db

# Load environment variables
load_dotenv(dotenv_path=os.path.join(os.path.dirname(__file__), ".env"))

ORACLE_PRIVATE_KEY = os.getenv("ORACLE_PRIVATE_KEY")
ALEO_NODE_URL = os.getenv("ALEO_NODE_URL")


def fetch_eth_staking_data():
    """
    Fetches ETH staking rate (total_staked / total_supply).
    Yields a percentage value (e.g., 28.5).
    """
    try:
        # In a real implementation, you would uses APIs like:
        # - Etherscan: Total Supply
        # - Beaconcha.in: Total Staked

        # For demonstration, we use sample values or a public API if available.
        # This is the "Snapshot" logic described by the user.

        # total_supply_url = "https://api.etherscan.io/api?module=stats&action=ethsupply&apikey=YOUR_API_KEY"
        # total_staked_url = "https://api.beaconcha.in/v1/validator/statistics"

        # Mock logic representing the "frozen value" at snapshot time
        # In production, this would be a real GET request.
        print("📸 Taking snapshot of ETH staking data...")

        # Sample data as of early 2026/late 2025 projection
        total_staked = 34500000  # Example: 34.5M ETH staked
        total_supply = 120000000  # Example: 120M ETH total supply

        staking_rate = (total_staked / total_supply) * 100
        print(f"📊 Snapshot Value: {staking_rate:.2f}%")

        return staking_rate
    except Exception as e:
        print(f"❌ Error fetching staking data: {e}")
        return None


def resolve_market(market_id, winning_option):
    """
    Executes the Aleo transaction to resolve the market using the Aleo SDK.
    """

    if not ORACLE_PRIVATE_KEY or not ALEO_NODE_URL:
        print("❌ Missing Oracle credentials in .env.")
        return False

    try:
        # 1. Setup Oracle Account and Query
        private_key = aleo.PrivateKey.from_string(ORACLE_PRIVATE_KEY)
        query = aleo.Query.rest(ALEO_NODE_URL)
        process = aleo.Process.load()

        # 2. Load the prediction program from the network
        program_id = "prediction.aleo"
        print(f"📡 Fetching program {program_id} from the network...")
        program_source = query.get_program(program_id)
        
        program = aleo.Program.from_string(program_source)
        process.add_program(program)

        # 3. Authorize Resolution
        print(
            f"🚀 Authorizing resolution for market {market_id} with option {winning_option}..."
        )
        function_name = aleo.Identifier.from_string("resolve_pool")

        # Inputs: market_id (field), winning_option (u64)
        inputs = [
            aleo.Value.from_string(market_id),
            aleo.Value.from_string(f"{winning_option}u64"),
        ]

        auth = process.authorize(private_key, program.id(), function_name, inputs)

        # 4. Execute and Prove
        print("🔧 Generating zero-knowledge proof...")
        (resp, trace) = process.execute(auth)
        trace.prepare(query)

        execution = trace.prove_execution(aleo.Locator(program.id(), function_name))
        execution_id = execution.execution_id()
        process.verify_execution(execution)

        # 5. Handle Fees (Public Fee)
        # For simplicity, we use a fixed fee or calculate it if costs are known
        # In a real environment, you'd check process.execution_cost(execution)
        fee_cost = 100_000  # Example fee in microcredits
        fee_auth = process.authorize_fee_public(
            private_key, fee_cost, execution_id, None
        )
        (_fee_resp, fee_trace) = process.execute(fee_auth)
        fee_trace.prepare(query)
        fee = fee_trace.prove_fee()
        process.verify_fee(fee, execution_id)

        # 6. Create and Broadcast Transaction
        transaction = aleo.Transaction.from_execution(execution, fee)
        print(f"📡 Broadcasting transaction {execution_id}...")

        # Use query to broadcast
        # Note: If Query doesn't have a direct broadcast, we'd use requests to POST to the node's /transaction/broadcast endpoint
        tx_json = transaction.to_json()
        response = requests.post(
            f"{ALEO_NODE_URL}/testnet3/transaction/broadcast", data=tx_json
        )

        if response.status_code == 200:
            print(f"✅ Transaction Broadcasted! ID: {response.text.strip()}")
            return True
        else:
            print(f"❌ Broadcast failed: {response.text}")
            return False

    except Exception as e:
        print(f"❌ SDK Error during resolution: {e}")
        return False


def main():
    db.init_db()
    print("🤖 Oracle Worker is running and monitoring pending markets...")

    while True:
        pending = db.get_pending_markets()
        current_time = int(time.time())

        for market_id, deadline, threshold, metric_type in pending:
            if current_time >= deadline:
                print(f"⏰ Deadline reached for market: {market_id}")

                value = fetch_eth_staking_data()
                if value is not None:
                    # Resolve: Option 1 (Yes) if value >= threshold, Option 2 (No) otherwise
                    winning_option = 1 if value >= threshold else 2

                    if resolve_market(market_id, winning_option):
                        db.mark_resolved(market_id)
                        print(
                            f"✅ Market {market_id} resolved as {'YES' if winning_option == 1 else 'NO'}"
                        )
                else:
                    print(
                        f"⚠️ Could not fetch data for market {market_id}, retrying next loop..."
                    )

        # Poll every 60 seconds
        time.sleep(60)


if __name__ == "__main__":
    main()

File: Oracle/src/worker.ts
import "@provablehq/sdk/testnet.js";
import {
    Account,
    AleoNetworkClient,
    NetworkRecordProvider,
    ProgramManager,
    AleoKeyProvider,
    initThreadPool,
} from "@provablehq/sdk";
import * as db from "./db.js";
import { registry } from "./metrics/registry.js";
import {
    ORACLE_PRIVATE_KEY,
    ALEO_NODE_URL,
    PROGRAM_ID,
    PROGRAM_SOURCE,
    CREATE_POOL_FEE,
    LOCK_POOL_FEE,
    RESOLVE_POOL_FEE,
} from "./config.js";

// Setup SDK
const account = new Account({ privateKey: ORACLE_PRIVATE_KEY });
const networkClient = new AleoNetworkClient(ALEO_NODE_URL);
const keyProvider = new AleoKeyProvider();
const recordProvider = new NetworkRecordProvider(account, networkClient);
keyProvider.useCache(true);

const programManager = new ProgramManager(
    ALEO_NODE_URL,
    keyProvider,
    recordProvider
);

programManager.setAccount(account);

const stringToField = (str: string): string => {
    const buffer = Buffer.from(str, "utf8");
    let hex = buffer.toString("hex");
    if (buffer.length > 31) hex = buffer.subarray(0, 31).toString("hex");
    const bigInt = BigInt("0x" + hex);
    return `${bigInt}field`;
};

const fieldToString = (field: string): string => {
    try {
        const bigIntStr = field.replace("field", "");
        const bigInt = BigInt(bigIntStr);
        let hex = bigInt.toString(16);
        if (hex.length % 2 !== 0) hex = "0" + hex;
        const buffer = Buffer.from(hex, "hex");
        return buffer.toString("utf8");
    } catch (e) {
        return "Unknown";
    }
};

async function syncOnChainData() {
    console.log("🔄 Syncing on-chain market data...");
    const allMarkets = await db.getAllMarkets();

    for (const market of allMarkets) {
        const { market_id } = market;
        try {
            // Use market_id directly if it's already a field, otherwise convert
            const marketIdField = market_id.endsWith('field') ? market_id : stringToField(market_id);

            // Query the 'pools' mapping in the program
            const poolData: any = await networkClient.getProgramMappingValue(
                PROGRAM_ID,
                "pools",
                marketIdField
            );

            if (poolData) {
                console.log(`📊 Fetched on-chain data for ${market_id}`);

                // Parse stats
                const totalStakedMatch = poolData.match(/total_staked:\s*(\d+)u64/);
                const optionAStakesMatch = poolData.match(/option_a_stakes:\s*(\d+)u64/);
                const optionBStakesMatch = poolData.match(/option_b_stakes:\s*(\d+)u64/);

                if (totalStakedMatch && optionAStakesMatch && optionBStakesMatch) {
                    const totalStaked = parseInt(totalStakedMatch[1]);
                    const optionAStakes = parseInt(optionAStakesMatch[1]);
                    const optionBStakes = parseInt(optionBStakesMatch[1]);

                    await db.updateMarketStats(market_id, totalStaked, optionAStakes, optionBStakes);
                }
            }
        } catch (e: any) {
            if (!e.message.includes("not found")) {
                console.error(`❌ Error syncing data for ${market_id}: ${e.message}`);
            }
        }
    }
}

// Check if a market exists on-chain
async function marketExistsOnChain(marketId: string): Promise<boolean> {
    try {
        const marketIdField = marketId.endsWith('field') ? marketId : stringToField(marketId);
        const poolData = await networkClient.getProgramMappingValue(
            PROGRAM_ID,
            "pools",
            marketIdField
        );
        return poolData !== null && poolData !== undefined;
    } catch (e: any) {
        // "not found" means market doesn't exist on-chain
        return false;
    }
}

// Create a market on the Aleo blockchain
async function createMarketOnChain(market: any): Promise<boolean> {
    if (!ORACLE_PRIVATE_KEY || !ALEO_NODE_URL) {
        console.error("❌ Missing Oracle credentials in .env.");
        return false;
    }

    try {
        const { market_id, title, description, deadline, option_a_label, option_b_label } = market;

        console.log(`🚀 Creating market on-chain: ${title} (${market_id})`);

        // Convert strings to field format for Aleo
        const titleField = stringToField(title || "Market");
        const descField = stringToField(description || "Prediction market");
        const optionAField = stringToField(option_a_label || "YES");
        const optionBField = stringToField(option_b_label || "NO");
        const deadlineU64 = `${deadline}u64`;

        // Inputs for create_pool: title, description, options[2], deadline
        const inputs = [
            titleField,
            descField,
            `[${optionAField}, ${optionBField}]`,
            deadlineU64
        ];

        const fee = CREATE_POOL_FEE / 1_000_000;

        const executionResponse = await programManager.execute({
            programName: PROGRAM_ID,
            functionName: "create_pool",
            priorityFee: fee,
            privateFee: false,
            inputs: inputs,
            program: PROGRAM_SOURCE,
            keySearchParams: { cacheKey: `${PROGRAM_ID}:create_pool` }
        });

        console.log(`✅ Create Pool Transaction Broadcasted! ID: ${executionResponse}`);
        return true;
    } catch (e: any) {
        console.error(`❌ SDK Error during market creation: ${e.message}`);
        return false;
    }
}

async function lockMarket(marketId: string): Promise<boolean> {
    if (!ORACLE_PRIVATE_KEY || !ALEO_NODE_URL) {
        console.error("❌ Missing Oracle credentials in .env.");
        return false;
    }

    try {
        console.log(`🔒 Authorizing lock for market ${marketId}...`);

        const marketIdField = stringToField(marketId);
        const inputs = [marketIdField];
        const fee = LOCK_POOL_FEE / 1_000_000;

        const executionResponse = await programManager.execute({
            programName: PROGRAM_ID,
            functionName: "lock_pool",
            priorityFee: fee,
            privateFee: false,
            inputs: inputs,
            program: PROGRAM_SOURCE,
            keySearchParams: { cacheKey: `${PROGRAM_ID}:lock_pool` }
        });

        console.log(`✅ Lock Transaction Broadcasted! ID: ${executionResponse}`);
        return true;
    } catch (e: any) {
        console.error(`❌ SDK Error during locking: ${e.message}`);
        return false;
    }
}

async function resolveMarket(marketId: string, winningOption: number): Promise<boolean> {
    if (!ORACLE_PRIVATE_KEY || !ALEO_NODE_URL) {
        console.error("❌ Missing Oracle credentials in .env.");
        return false;
    }

    try {
        console.log(
            `🚀 Authorizing resolution for market ${marketId} with option ${winningOption}...`
        );

        const marketIdField = stringToField(marketId);
        const inputs = [marketIdField, `${winningOption}u64`];
        const fee = RESOLVE_POOL_FEE / 1_000_000;

        const executionResponse = await programManager.execute({
            programName: PROGRAM_ID,
            functionName: "resolve_pool",
            priorityFee: fee,
            privateFee: false,
            inputs: inputs,
            program: PROGRAM_SOURCE,
            keySearchParams: { cacheKey: `${PROGRAM_ID}:resolve_pool` }
        });

        console.log(`✅ Transaction Broadcasted! ID: ${executionResponse}`);
        return true;
    } catch (e: any) {
        console.error(`❌ SDK Error during resolution: ${e.message}`);
        return false;
    }
}

export async function startWorker() {
    await initThreadPool();
    await db.initDb();
    console.log("🤖 Oracle Worker is running and monitoring pending markets...");

    setInterval(async () => {
        try {
            const currentTime = Math.floor(Date.now() / 1000);

            // 0. Sync on-chain stats (TVL, stakes)
            await syncOnChainData();

            // 1. Check for markets that need to be created on-chain
            const marketsNotOnChain = await db.getMarketsNotOnChain();
            for (const market of marketsNotOnChain) {
                const { market_id, title } = market;

                console.log(`📝 Market "${title}" not on-chain. Creating...`);
                const success = await createMarketOnChain(market);
                if (success) {
                    // Mark as on-chain in database (transaction submitted)
                    await db.markOnChain(market_id);
                    console.log(`✅ Market "${title}" creation transaction submitted.`);
                }
                // Add delay to avoid rate limiting
                await new Promise(resolve => setTimeout(resolve, 5000));
            }

            // 2. Handle Pending -> Locked
            const pending = await db.getPendingMarkets();
            for (const market of pending) {
                const { market_id, deadline } = market;
                if (currentTime >= deadline) {
                    console.log(`⏰ Deadline reached for market: ${market_id}. Locking...`);
                    const success = await lockMarket(market_id);
                    if (success) {
                        await db.markLocked(market_id);
                    }
                }
            }

            // 3. Handle Locked -> Resolved
            const locked = await db.getLockedMarkets();
            for (const market of locked) {
                const { market_id, threshold, metric_type } = market;

                console.log(`🔎 Resolution check for locked market: ${market_id} (Metric: ${metric_type})`);

                const handler = registry.getMetric(metric_type);
                if (!handler) {
                    console.error(`❌ No handler found for metric type: ${metric_type}`);
                    continue;
                }

                const value = await handler.fetchValue();
                if (value !== null) {
                    const winningOption = value >= threshold ? 1 : 2;

                    const success = await resolveMarket(market_id, winningOption);
                    if (success) {
                        await db.markResolved(market_id);
                        console.log(
                            `✅ Market ${market_id} resolved as ${winningOption === 1 ? "YES" : "NO"}`
                        );
                    }
                } else {
                    console.log(`⚠️ Could not fetch data for market ${market_id}, retrying next loop...`);
                }
            }
        } catch (e: any) {
            console.error("Error in worker loop:", e.message);
        }
    }, 60000); // 60 seconds
}

// Start worker if run directly
import { fileURLToPath } from "url";
if (process.argv[1] === fileURLToPath(import.meta.url)) {
    startWorker().catch(console.error);
}
